paasta_tools/subscribe_mesos_master.py

```diff
@@ -0,0 +1,247 @@
+            return f"Unknown: TASK_KILLED but unrecognized status.healthy={healthy!r}"
+    elif state == "TASK_FAILED":
+        if message == "Container exited with status 137":
+            return f"Probable OOM: {message}"
+        elif message is None:
+            return f"Unknown: TASK_FAILED but status.message is None. This seems to happen when a task exits very quickly on startup. Mesos usually follows up with a corrected message shortly."
+        elif message.startswith("Container exited with status "):
+            return f"TASK_FAILED: {message}"
+        else:
+            return f"Unknown: TASK_FAILED but unrecognized status.message"
+    elif state == "TASK_FINISHED":
+        if message is None:
+            return f"Unknown: TASK_FINISHED but status.message is None. This seems to happen when a task exits very quickly on startup. Mesos usually follows up with a corrected message shortly."
+        return f"TASK_FINISHED: {message}"
+    else:
+        return f"Unknown: unrecognized state"
+
+
+class MesosEventSubscriber:
+    def __init__(self, cluster):
+        self.framework_id_to_name = {}
+        self.framework_id_task_id_to_git_sha = {}
+        self.cluster = cluster
+
+    async def subscribe(self):
+        # This connection should live ~forever, so disable some timeouts.
+        timeout = aiohttp.ClientTimeout(
+            total=None, sock_read=None, connect=30, sock_connect=30,
+        )
+        async with aiohttp.ClientSession(timeout=timeout) as session:
+            payload = '{"type":"SUBSCRIBE"}'
+            master_host_port = mesos_tools.find_mesos_leader(cluster=self.cluster)
+            async with session.post(
+                f"http://{master_host_port}/api/v1",
+                data=payload,
+                # allow_redirects=True,
+                headers={"Content-Type": "application/json"},
+                timeout=timeout,
+            ) as resp:
+                while True:
+                    _size = await resp.content.readline()
+                    if not _size:
+                        break
+                    size = int(_size)
+                    record = await resp.content.readexactly(size)
+                    yield json.loads(record)
+
+    def determine_service_instance(self, framework_name, status):
+        executor_id = status.get("executor_id", {}).get("value")
+        task_id = status.get("task_id", {}).get("value")
+
+        if framework_name.startswith("marathon"):
+            return marathon_tools.parse_service_instance_from_executor_id(
+                executor_id or task_id
+            )
+        elif framework_name.startswith("tron"):
+            return tron_tools.decompose_executor_id(executor_id)[:2]
+        elif framework_name.startswith("paasta-remote "):
+            # sorta gross, but it's the same format.
+            return marathon_tools.parse_service_instance_from_executor_id(executor_id)
+        else:
+            raise UnknownFrameworkError(
```

Comment:
I looks like this is fatal?

Can you catch this later on so that this script wont crash when a jolt run happens or spark or whatever?

Reply:
Oh, you have the skip thing in another function.

Hmm. Just kinda concerns me that we assume we know all the framework names. If like, chronos started randomly today this script would crash?